runtime/src/bank/builtin_programs.rs

```diff
@@ -62,3 +62,399 @@ mod tests {
+    #[test_case(TestPrototype::Builtin(&BUILTINS[8]); "address_lookup_table")]
+    #[test_case(TestPrototype::Stateless(&STATELESS_BUILTINS[0]); "feature_gate")]
+    fn test_core_bpf_migration(prototype: TestPrototype) {
+        let (genesis_config, mint_keypair) = create_genesis_config(1_000_000 * LAMPORTS_PER_SOL);
+        let mut root_bank = Bank::new_for_tests(&genesis_config);
+
+        let (builtin_id, config) = prototype.deconstruct();
+        let feature_id = &config.feature_id;
+        let source_program_id = &config.source_program_id;
+
+        // Add the feature to the bank's inactive feature set.
+        let mut feature_set = FeatureSet::all_enabled();
+        feature_set.inactive.insert(*feature_id);
+        root_bank.feature_set = Arc::new(feature_set);
+
+        // Initialize the accounts for the source program.
+        let test_context = TestContext::new(&root_bank, builtin_id, source_program_id);
+
+        let (bank, bank_forks) = root_bank.wrap_with_bank_forks_for_tests();
+
+        // Advance one slot so that the source program becomes effective in the
+        // program cache.
+        goto_end_of_slot(bank.clone());
+        let bank = new_from_parent_with_fork_next_slot(bank, bank_forks.as_ref());
+
+        // Successfully invoke the source program.
+        bank.process_transaction(&Transaction::new(
+            &vec![&mint_keypair],
+            Message::new(
+                &[Instruction::new_with_bytes(
+                    *source_program_id,
+                    &[],
+                    Vec::new(),
+                )],
+                Some(&mint_keypair.pubkey()),
+            ),
+            bank.last_blockhash(),
+        ))
+        .unwrap();
+
+        // Advance to the next epoch without activating the feature.
+        let bank = new_bank_from_parent_with_bank_forks(&bank_forks, bank, &Pubkey::default(), 33);
+
+        // Assert the feature was not activated and the program was not
+        // migrated.
+        assert!(!bank.feature_set.is_active(feature_id));
+        assert!(bank.get_account(source_program_id).is_some());
+
+        // Store the account to activate the feature.
+        bank.store_account_and_update_capitalization(
+            feature_id,
+            &feature::create_account(&Feature::default(), 42),
+        );
+
+        // Advance the bank to cross the epoch boundary and activate the
+        // feature.
+        goto_end_of_slot(bank.clone());
+        let bank = new_bank_from_parent_with_bank_forks(&bank_forks, bank, &Pubkey::default(), 96);
+
+        // Run the post-migration program checks.
+        assert!(bank.feature_set.is_active(feature_id));
+        test_context.run_program_checks_post_migration(&bank, 96);
+
+        // Advance one slot so that the new target program becomes effective in
+        // the program cache.
+        goto_end_of_slot(bank.clone());
+        let bank = new_from_parent_with_fork_next_slot(bank, bank_forks.as_ref());
```

Comment:
The migration is conducted on feature activation during epoch rollover, so it has to be the 0th slot of the epoch where the program is deployed. Unless I should somehow move out of `apply_builtin_program_feature_transitions`?

Reply:
Did the SIMD specify whether the deployment slot or the effective slot was at the boundary?

Reply:
No, it did not. It just said we'd deploy a program as the source to migrate into the address of the builtin on epoch rollover. What are you thinking?

Reply:
After discussions offline, marking this conversation resolved, since it seems having a one-slot delayed visibility is acceptable.
https://discord.com/channels/428295358100013066/1204657211402489876/1229478107682181256